### PR TITLE
Docathon: Module level docs for Pulse

### LINKED ISSUE #3406: qiskit.pulse module and underlying functionality needs documentation.

### PR DESCRIPTION
**Pull request by lcapelluto**
### Summary
Update the pulse module init file for documentation. This includes adding autosummary references to important parts of the module that were previously unreachable, and providing introductory summaries where appropriate

This also closes #3406.

### Details and comments

 - [x] One bug that @SooluThomas is helping me with: can't seem to link to the `commands` module! (made an issue)

~Also: I will wait on #3838 to merge first, since I anticipate merge conflicts~

### CODE REVIEW

File: qiskit/pulse/__init__.py [hunk 1]
```diff
@@ -13,56 +13,77 @@
...
+Qiskit-Pulse is a pulse-level quantum programming kit. This lower level of programming offers the
+user more control than programming with :py:class:`~qiskit.circuit.QuantumCircuit` s.
 
-   DriveChannel
-   MeasureChannel
-   AcquireChannel
-   ControlChannel
-   RegisterSlot
-   MemorySlot
+Extracting the greatest performance out from quantum hardware requires real-time pulse-level
```
Comment (DanPuzzuoli) [inline qiskit/pulse/__init__.py]: I see why 'out' is in there but I think it flows better without it.

```suggestion
Extracting the greatest performance from quantum hardware requires real-time pulse-level
```

File: qiskit/pulse/__init__.py [hunk 2]
```diff
@@ -13,56 +13,77 @@
...
+
+Channels (:mod:`~qiskit.pulse.channels`)
+========================================
+
+Pulse is meant to be agnostic to the underlying hardware implementation, while still allowing
+low-level control. Therefore, our signal channels are  *virtual* hardware channels. The backend
+which executes our programs is responsible for mapping these virtual channels to the proper
+physical channel within the quantum control hardware.
+
+Channels are characterized by their type and their index. See each of the below to learn more.
```
Comment (DanPuzzuoli) [inline qiskit/pulse/__init__.py]: I think something like 'channel type', 'channel class', or even 'channel-type class' (lol) will make this last sentence read better.

```suggestion
Channels are characterized by their type and their index. See each channel type below to learn more.
```

File: qiskit/pulse/__init__.py (hunk 1 above)
Comment (lcapelluto) [reply qiskit/pulse/__init__.py]: agree

**Comment by lcapelluto** [comment]
Here is the follow up bug report: https://github.com/Qiskit/qiskit-terra/issues/3859

**Pull request merged**